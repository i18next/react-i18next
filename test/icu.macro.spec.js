import pluginTester from 'babel-plugin-tester';
import plugin from 'babel-plugin-macros';

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename, parserOpts: { plugins: ['jsx'] } },
  tests: [
    `
      import { Trans } from '../icu.macro'

      const x = <Trans>Welcome, { name }!</Trans>
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans>Welcome, <strong>{ name }</strong>!</Trans>
    `,

    `
      import { Trans } from '../icu.macro'
      import { useTranslation } from 'react-i18next'

      const x = <Trans>Trainers: { trainersCount, number }</Trans>
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans>Trainers: <strong>{ trainersCount, number }</strong>!</Trans>
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans>Caught on { catchDate, date, short }</Trans>
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans>Caught on <strong>{ catchDate, date, short }</strong>!</Trans>
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans defaults="Trainers: { trainersCount, number }" />
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans i18nKey="trainersWithDefaults" defaults="Trainers: <strong>{ trainersCount, number }</strong>!" />
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans i18nKey="caughtWithDefaults" defaults="Caught on { catchDate, date, short }" />
    `,

    `
      import { Trans } from '../icu.macro'

      const x = <Trans defaults="Caught on <strong>{ catchDate, date, short }</strong>!" />
    `,

    `
      import { Trans } from '../icu.macro'
      const Link = ({to, children}) => (<a href={to}>{children}</a>)
      
      const x = <Trans defaults="Caught on <Link to='/dest'>{ catchDate, date, short }</Link>!" values={{catchDate: Date.now()}}>This should be overridden by defaults</Trans>
    `,

    `
      import { Trans } from '../icu.macro'
      
      const x = <Trans i18nKey="trainersWithDefaults" values={{trainersCount}} defaults="Trainers: <strong>{ trainersCount, number }</strong>!" components={[]} />
    `,

    `
      import { Select } from '../icu.macro'

      const x = <Select
        i18nKey="selectExample"
        switch={gender}
        male="He avoids bugs."
        female="She avoids bugs."
        other="They avoid bugs."
      />
    `,

    `
      import { Select } from '../icu.macro'

      const x = <Select
        switch={gender}
        male={<Trans><strong>He</strong> avoids bugs.</Trans>}
        female={<Trans><strong>She</strong> avoids bugs.</Trans>}
        other={<Trans><strong>They</strong> avoid bugs.</Trans>}
      />
    `,

    `
      import { Plural } from '../icu.macro'

      const x = <Plural
        count={itemsCount1}
        $0="There are no items."
        one="There is # item."
        other="There are # items."
      />
    `,

    `
      import { Plural } from '../icu.macro'

      const x = <Plural
        i18nKey="testKey"
        count={itemsCount3}
        $0={<Trans>There is <strong>no</strong> item.</Trans>}
        one={<Trans>There is <strong>#</strong> item.</Trans>}
        other={<Trans>There are <strong>#</strong> items.</Trans>}
      />
    `,

    `
      import { Plural } from '../icu.macro'

      const x = <Plural
        i18nKey="testKey"
        count={itemsCount3}
        values={{location: 'table'}}
        $0={<Trans>There is <strong>no</strong> item on the <i>{location}</i>.</Trans>}
        one={<Trans>There is <strong>#</strong> item on the <i>{location}</i>.</Trans>}
        other={<Trans>There are <strong>#</strong> items on the <i>{location}</i>.</Trans>}
      />
    `,

    `
      import { SelectOrdinal } from '../icu.macro'

      const x = <SelectOrdinal
        count={position}
        zero="You are #th in line"
        one="You are #st in line"
        two="You are #nd in line"
        few="You are #rd in line"
        many="You are #th in line"
        other="You are #th in line"
        $7="You are in the lucky #th place in line"
      />
    `,

    `
      import { SelectOrdinal } from '../icu.macro'

      const x = <SelectOrdinal
        i18nKey="testKey"
        count={position}
        one={<Trans>You are <strong>#st in line</strong></Trans>}
        two={<Trans>You are <strong>#nd in line</strong></Trans>}
        few={<Trans>You are <strong>#rd in line</strong></Trans>}
        other={<Trans>You are <strong>#th in line</strong></Trans>}
        $0={<Trans>You are <strong>#th in line</strong></Trans>}
      />
    `,

    `
      import React from 'react'
      import { useTranslation } from 'react-i18next'
      import { Plural, Select, SelectOrdinal, Trans } from '../icu.macro'
      const Link = ({to, children}) => (<a href={to}>{children}</a>)
      
      export default function TestPage({count = 1}) {
        const [t] = useTranslation()
        const catchDate = Date.now()
        const completion = 0.75
        const gender = Math.random() < 0.5 ? 'female' : 'male'
        return (
          <>
            {t('sample.text', 'Some sample text with {word} {gender} {count, number} {catchDate, date} {completion, number, percent}', {word: 'interpolation', gender, count, catchDate, completion})}
            <Plural i18nKey="plural"
              count={count}
              values={{linkPath: "/item/" + count}}
              $0={<Trans><Link to='/cart'>Your cart</Link> is <strong>empty</strong>.</Trans>}
              one={<Trans>You have <strong># item</strong> in <Link to='/cart'>your cart</Link>.</Trans>}
              other={<Trans>You have <strong># items</strong> in <Link to='/cart'>your cart</Link>.</Trans>}
            />
            <Select
              i18nKey="select"
              switch={gender}
              female={<Trans>These are <Link to='/items'>her items</Link></Trans>}
              male={<Trans>These are <Link to='/items'>his items</Link></Trans>}
              other={<Trans>These are <Link to='/items'>their items</Link></Trans>}
            />
            <SelectOrdinal i18nKey="ordinal"
              count={itemIndex+1}
              values={{linkPath: "/item/" + itemIndex}}
              one={<Trans>Your <Link to={linkPath}><strong>#st</strong> item</Link></Trans>}
              two={<Trans>Your <Link to={linkPath}><strong>#nd</strong> item</Link></Trans>}
              few={<Trans>Your <Link to={linkPath}><strong>#rd</strong> item</Link></Trans>}
              other={<Trans>Your <Link to={linkPath}><strong>#th</strong> item</Link></Trans>}
            />
            <Trans i18nKey="percent" defaults="You&apos;ve completed <Link to='/tasks'>{completion, number, percent} of your tasks</Link>."/>
            <Trans i18nKey="date" defaults="Caught on <Link to='/dest'>{ catchDate, date, short }</Link>!"/>
            <SelectOrdinal
              i18nKey="ordinal.prettier"
              count={count}
              values={{ linkPath: \`/item/\${count}\`, type: 'item', prop }}
              one={
                <Trans>
                  Your{' '}
                  <Link to={linkPath}>
                    <strong>#st</strong> {type}
                  </Link>
                </Trans>
              }
              two={
                <Trans>
                  Your{' '}
                  <Link to={linkPath}>
                    <strong>#nd</strong> {type}
                  </Link>
                </Trans>
              }
              few={
                <Trans>
                  Your{' '}
                  <Link to={linkPath}>
                    <strong>#rd</strong> {type}
                  </Link>
                </Trans>
              }
              other={
                <Trans>
                  Your{' '}
                  <Link to={linkPath}>
                    <strong>#th</strong> {type}
                  </Link>
                </Trans>
              }
            />
            <Select
              i18nKey="select.expr.prettier"
              switch={\`\${gender}Person\`}
              values={{ linkPath: \`/users/\${number}\`, type: 'bugs' }}
              malePerson={
                <Trans>
                  <Link to={linkPath}>
                    <strong>He</strong>
                  </Link>{' '}
                  avoids {type}.
                </Trans>
              }
              femalePerson={
                <Trans>
                  <Link to={linkPath}>
                    <strong>She</strong>
                  </Link>{' '}
                  avoids {type}.
                </Trans>
              }
              other={
                <Trans>
                  <Link to={linkPath}>
                    <strong>They</strong>
                  </Link>{' '}
                  avoid {type}.
                </Trans>
              }
            />
          </>
        )
      }
    `,
    `
      import React from "react"
      import { Trans, number, date, time, plural, select, selectOrdinal } from "../icu.macro";

      function Component({ children, style }) {
        return <div style={style}>{children}</div>
      }

      const count = 2;
      const numbers = 34;
      const selectInput = "thing"
      const now = new Date()
      const x = (
        <Trans i18nKey="key">
          <strong>exciting!</strong>
          {plural\`\${count},
          =0 { hi there \${<strong>friend</strong>} }
          other { woweee even supports nested \${number\`\${numbers}\`} } \`} and
          {select\`\${selectInput},
           thing { another nested \${<Component style={{ color: "red" }}>
             with regular text and a date: <pre>{date\`\${now}\`}</pre>
           </Component>}} \`}
        </Trans>
      );
    `,
    {
      code: `
        import React from "react"
        import { Trans, number } from "../icu.macro";

        const count = 2;
        const outside = number\`\${count}\`;
      `,
      snapshot: false,
      error: /"number``" can only be used inside <Trans> in "[^"]+" on line 5/,
    },
    {
      code: `
        import React from "react"
        import { Trans, date } from "../icu.macro";

        const d = new Date;
        const outside = date\`\${d}\`;
      `,
      snapshot: false,
      error: /"date``" can only be used inside <Trans> in "[^"]+" on line 5/,
    },
    {
      code: `
        import React from "react"
        import { Trans, time } from "../icu.macro";

        const d = new Date;
        const outside = time\`\${d}\`;
      `,
      snapshot: false,
      error: /"time``" can only be used inside <Trans> in "[^"]+" on line 5/,
    },
    {
      code: `
        import React from "react"
        import { Trans, select } from "../icu.macro";

        const d = "f";
        const outside = select\`\${d}, f { chose f } other { chose something else }\`;
      `,
      snapshot: false,
      error: /"select``" can only be used inside <Trans> in "[^"]+" on line 5/,
    },
    {
      code: `
        import React from "react"
        import { Trans, selectOrdinal } from "../icu.macro";

        const d = 1;
        const outside = selectOrdinal\`\${d}, =0 { # } other { chose # }\`;
      `,
      snapshot: false,
      error: /"selectOrdinal``" can only be used inside <Trans> in "[^"]+" on line 5/,
    },
    {
      code: `
        import React from "react"
        import { Trans, plural } from "../icu.macro";

        const d = 1;
        const outside = plural\`\${d}, =0 { # } other { chose # }\`;
      `,
      snapshot: false,
      error: /"plural``" can only be used inside <Trans> in "[^"]+" on line 5/,
    },
    {
      code: `
        import React from "react"
        import { Trans, plural } from "../icu.macro";

        const d = 1;
        const x = (
          <Trans i18nKey="key">
           this will {fail\`hard\`}
          </Trans>
        );
      `,
      snapshot: false,
      error: /Unsupported tagged template literal "fail", must be one of date, time, number, plural, select, selectOrdinal in "[^"]+" on line 7/,
    },
    {
      code: `
        import React from "react"
        import { Trans, date } from "../icu.macro";

        const d = 1;
        const x = (
          <Trans i18nKey="key">
           this will {date\`fail\`}
          </Trans>
        );
      `,
      snapshot: false,
      error: /date argument must be interpolated in "date``" in "[^"]+" on line 7/,
    },
    {
      code: `
        import React from "react"
        import { Trans, date } from "../icu.macro";

        const tooLate = new Date();
        const x = (
          <Trans i18nKey="key">
           this will {date\`fail\${tooLate}\`}
          </Trans>
        );
      `,
      snapshot: false,
      error: /date argument must be interpolated at the beginning of "date``" in "[^"]+" on line 7/,
    },
    {
      code: `
        import React from "react"
        import { Trans, number } from "../icu.macro";

        const tooLate = funcThatReturnsNumberOrUndefined;
        const x = (
          <Trans i18nKey="key">
           this will {number\`\${tooLate || 0}\`}
          </Trans>
        );
      `,
      snapshot: false,
      error: /Must pass a variable, not an expression to "number``" in "[^"]+" on line 7/,
    },
  ],
});
