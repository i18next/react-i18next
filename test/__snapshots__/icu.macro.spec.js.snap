// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros 1. macros: 1. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans>Welcome, { name }!</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Welcome, {name}!\\" components={[]} values={{
  name
}} />;
"
`;

exports[`macros 2. macros: 2. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans>Welcome, <strong>{ name }</strong>!</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Welcome, <0>{name}</0>!\\" components={[<strong>{name}</strong>]} values={{
  name
}} />;
"
`;

exports[`macros 3. macros: 3. macros 1`] = `
"
import { Trans } from '../icu.macro'
import { useTranslation } from 'react-i18next'

const x = <Trans>Trainers: { trainersCount, number }</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { useTranslation, Trans } from 'react-i18next';
const x = <Trans defaults=\\"Trainers: {trainersCount, number}\\" components={[]} values={{
  trainersCount
}} />;
"
`;

exports[`macros 4. macros: 4. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans>Trainers: <strong>{ trainersCount, number }</strong>!</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Trainers: <0>{trainersCount, number}</0>!\\" components={[<strong>{(trainersCount)}</strong>]} values={{
  trainersCount
}} />;
"
`;

exports[`macros 5. macros: 5. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans>Caught on { catchDate, date, short }</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Caught on {catchDate, date, short}\\" components={[]} values={{
  catchDate
}} />;
"
`;

exports[`macros 6. macros: 6. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans>Caught on <strong>{ catchDate, date, short }</strong>!</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Caught on <0>{catchDate, date, short}</0>!\\" components={[<strong>{(catchDate)}</strong>]} values={{
  catchDate
}} />;
"
`;

exports[`macros 7. macros: 7. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans defaults=\\"Trainers: { trainersCount, number }\\" />

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Trainers: {trainersCount, number}\\" components={[]} values={{
  trainersCount
}} />;
"
`;

exports[`macros 8. macros: 8. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans i18nKey=\\"trainersWithDefaults\\" defaults=\\"Trainers: <strong>{ trainersCount, number }</strong>!\\" />

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"trainersWithDefaults\\" defaults=\\"Trainers: <0>{trainersCount, number}</0>!\\" components={[<strong>{(trainersCount)}</strong>]} values={{
  trainersCount
}} />;
"
`;

exports[`macros 9. macros: 9. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans i18nKey=\\"caughtWithDefaults\\" defaults=\\"Caught on { catchDate, date, short }\\" />

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"caughtWithDefaults\\" defaults=\\"Caught on {catchDate, date, short}\\" components={[]} values={{
  catchDate
}} />;
"
`;

exports[`macros 10. macros: 10. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans defaults=\\"Caught on <strong>{ catchDate, date, short }</strong>!\\" />

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"Caught on <0>{catchDate, date, short}</0>!\\" components={[<strong>{(catchDate)}</strong>]} values={{
  catchDate
}} />;
"
`;

exports[`macros 11. macros: 11. macros 1`] = `
"
import { Trans } from '../icu.macro'
const Link = ({to, children}) => (<a href={to}>{children}</a>)

const x = <Trans defaults=\\"Caught on <Link to='/dest'>{ catchDate, date, short }</Link>!\\" values={{catchDate: Date.now()}}>This should be overridden by defaults</Trans>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";

const Link = ({
  to,
  children
}) => <a href={to}>{children}</a>;

const x = <Trans values={{
  catchDate: Date.now()
}} defaults=\\"Caught on <0>{catchDate, date, short}</0>!\\" components={[<Link to='/dest'>{(catchDate)}</Link>]} />;
"
`;

exports[`macros 12. macros: 12. macros 1`] = `
"
import { Trans } from '../icu.macro'

const x = <Trans i18nKey=\\"trainersWithDefaults\\" values={{trainersCount}} defaults=\\"Trainers: <strong>{ trainersCount, number }</strong>!\\" components={[]} />

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"trainersWithDefaults\\" values={{
  trainersCount
}} defaults=\\"Trainers: <strong>{ trainersCount, number }</strong>!\\" components={[]} />;
"
`;

exports[`macros 13. macros: 13. macros 1`] = `
"
import { Select } from '../icu.macro'

const x = <Select
  i18nKey=\\"selectExample\\"
  switch={gender}
  male=\\"He avoids bugs.\\"
  female=\\"She avoids bugs.\\"
  other=\\"They avoid bugs.\\"
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"selectExample\\" defaults=\\"{gender, select,  male {He avoids bugs.} female {She avoids bugs.} other {They avoid bugs.}}\\" components={[]} values={{
  gender
}} />;
"
`;

exports[`macros 14. macros: 14. macros 1`] = `
"
import { Select } from '../icu.macro'

const x = <Select
  switch={gender}
  male={<Trans><strong>He</strong> avoids bugs.</Trans>}
  female={<Trans><strong>She</strong> avoids bugs.</Trans>}
  other={<Trans><strong>They</strong> avoid bugs.</Trans>}
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"{gender, select,  male {<0>He</0> avoids bugs.} female {<1>She</1> avoids bugs.} other {<2>They</2> avoid bugs.}}\\" components={[<strong>He</strong>, <strong>She</strong>, <strong>They</strong>]} values={{
  gender
}} />;
"
`;

exports[`macros 15. macros: 15. macros 1`] = `
"
import { Plural } from '../icu.macro'

const x = <Plural
  count={itemsCount1}
  $0=\\"There are no items.\\"
  one=\\"There is # item.\\"
  other=\\"There are # items.\\"
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"{itemsCount1, plural,  =0 {There are no items.} one {There is # item.} other {There are # items.}}\\" components={[]} values={{
  itemsCount1
}} />;
"
`;

exports[`macros 16. macros: 16. macros 1`] = `
"
import { Plural } from '../icu.macro'

const x = <Plural
  i18nKey=\\"testKey\\"
  count={itemsCount3}
  $0={<Trans>There is <strong>no</strong> item.</Trans>}
  one={<Trans>There is <strong>#</strong> item.</Trans>}
  other={<Trans>There are <strong>#</strong> items.</Trans>}
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"testKey\\" defaults=\\"{itemsCount3, plural,  =0 {There is <0>no</0> item.} one {There is <1>#</1> item.} other {There are <2>#</2> items.}}\\" components={[<strong>no</strong>, <strong>#</strong>, <strong>#</strong>]} values={{
  itemsCount3
}} />;
"
`;

exports[`macros 17. macros: 17. macros 1`] = `
"
import { Plural } from '../icu.macro'

const x = <Plural
  i18nKey=\\"testKey\\"
  count={itemsCount3}
  values={{location: 'table'}}
  $0={<Trans>There is <strong>no</strong> item on the <i>{location}</i>.</Trans>}
  one={<Trans>There is <strong>#</strong> item on the <i>{location}</i>.</Trans>}
  other={<Trans>There are <strong>#</strong> items on the <i>{location}</i>.</Trans>}
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"testKey\\" defaults=\\"{itemsCount3, plural,  =0 {There is <0>no</0> item on the <1>{location}</1>.} one {There is <2>#</2> item on the <3>{location}</3>.} other {There are <4>#</4> items on the <5>{location}</5>.}}\\" components={[<strong>no</strong>, <i>{location}</i>, <strong>#</strong>, <i>{location}</i>, <strong>#</strong>, <i>{location}</i>]} values={{
  itemsCount3,
  location: 'table'
}} />;
"
`;

exports[`macros 18. macros: 18. macros 1`] = `
"
import { SelectOrdinal } from '../icu.macro'

const x = <SelectOrdinal
  count={position}
  zero=\\"You are #th in line\\"
  one=\\"You are #st in line\\"
  two=\\"You are #nd in line\\"
  few=\\"You are #rd in line\\"
  many=\\"You are #th in line\\"
  other=\\"You are #th in line\\"
  $7=\\"You are in the lucky #th place in line\\"
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans defaults=\\"{position, selectordinal,  zero {You are #th in line} one {You are #st in line} two {You are #nd in line} few {You are #rd in line} many {You are #th in line} other {You are #th in line} =7 {You are in the lucky #th place in line}}\\" components={[]} values={{
  position
}} />;
"
`;

exports[`macros 19. macros: 19. macros 1`] = `
"
import { SelectOrdinal } from '../icu.macro'

const x = <SelectOrdinal
  i18nKey=\\"testKey\\"
  count={position}
  one={<Trans>You are <strong>#st in line</strong></Trans>}
  two={<Trans>You are <strong>#nd in line</strong></Trans>}
  few={<Trans>You are <strong>#rd in line</strong></Trans>}
  other={<Trans>You are <strong>#th in line</strong></Trans>}
  $0={<Trans>You are <strong>#th in line</strong></Trans>}
/>

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans } from \\"react-i18next\\";
const x = <Trans i18nKey=\\"testKey\\" defaults=\\"{position, selectordinal,  one {You are <0>#st in line</0>} two {You are <1>#nd in line</1>} few {You are <2>#rd in line</2>} other {You are <3>#th in line</3>} =0 {You are <4>#th in line</4>}}\\" components={[<strong>#st in line</strong>, <strong>#nd in line</strong>, <strong>#rd in line</strong>, <strong>#th in line</strong>, <strong>#th in line</strong>]} values={{
  position
}} />;
"
`;

exports[`macros 20. macros: 20. macros 1`] = `
"
import React from 'react'
import { useTranslation } from 'react-i18next'
import { Plural, Select, SelectOrdinal, Trans } from '../icu.macro'
const Link = ({to, children}) => (<a href={to}>{children}</a>)

export default function TestPage({count = 1}) {
  const [t] = useTranslation()
  const catchDate = Date.now()
  const completion = 0.75
  const gender = Math.random() < 0.5 ? 'female' : 'male'
  return (
    <>
      {t('sample.text', 'Some sample text with {word} {gender} {count, number} {catchDate, date} {completion, number, percent}', {word: 'interpolation', gender, count, catchDate, completion})}
      <Plural i18nKey=\\"plural\\"
        count={count}
        values={{linkPath: \\"/item/\\" + count}}
        $0={<Trans><Link to='/cart'>Your cart</Link> is <strong>empty</strong>.</Trans>}
        one={<Trans>You have <strong># item</strong> in <Link to='/cart'>your cart</Link>.</Trans>}
        other={<Trans>You have <strong># items</strong> in <Link to='/cart'>your cart</Link>.</Trans>}
      />
      <Select
        i18nKey=\\"select\\"
        switch={gender}
        female={<Trans>These are <Link to='/items'>her items</Link></Trans>}
        male={<Trans>These are <Link to='/items'>his items</Link></Trans>}
        other={<Trans>These are <Link to='/items'>their items</Link></Trans>}
      />
      <SelectOrdinal i18nKey=\\"ordinal\\"
        count={itemIndex+1}
        values={{linkPath: \\"/item/\\" + itemIndex}}
        one={<Trans>Your <Link to={linkPath}><strong>#st</strong> item</Link></Trans>}
        two={<Trans>Your <Link to={linkPath}><strong>#nd</strong> item</Link></Trans>}
        few={<Trans>Your <Link to={linkPath}><strong>#rd</strong> item</Link></Trans>}
        other={<Trans>Your <Link to={linkPath}><strong>#th</strong> item</Link></Trans>}
      />
      <Trans i18nKey=\\"percent\\" defaults=\\"You&apos;ve completed <Link to='/tasks'>{completion, number, percent} of your tasks</Link>.\\"/>
      <Trans i18nKey=\\"date\\" defaults=\\"Caught on <Link to='/dest'>{ catchDate, date, short }</Link>!\\"/>
      <SelectOrdinal
        i18nKey=\\"ordinal.prettier\\"
        count={count}
        values={{ linkPath: \`/item/\${count}\`, type: 'item', prop }}
        one={
          <Trans>
            Your{' '}
            <Link to={linkPath}>
              <strong>#st</strong> {type}
            </Link>
          </Trans>
        }
        two={
          <Trans>
            Your{' '}
            <Link to={linkPath}>
              <strong>#nd</strong> {type}
            </Link>
          </Trans>
        }
        few={
          <Trans>
            Your{' '}
            <Link to={linkPath}>
              <strong>#rd</strong> {type}
            </Link>
          </Trans>
        }
        other={
          <Trans>
            Your{' '}
            <Link to={linkPath}>
              <strong>#th</strong> {type}
            </Link>
          </Trans>
        }
      />
      <Select
        i18nKey=\\"select.expr.prettier\\"
        switch={\`\${gender}Person\`}
        values={{ linkPath: \`/users/\${number}\`, type: 'bugs' }}
        malePerson={
          <Trans>
            <Link to={linkPath}>
              <strong>He</strong>
            </Link>{' '}
            avoids {type}.
          </Trans>
        }
        femalePerson={
          <Trans>
            <Link to={linkPath}>
              <strong>She</strong>
            </Link>{' '}
            avoids {type}.
          </Trans>
        }
        other={
          <Trans>
            <Link to={linkPath}>
              <strong>They</strong>
            </Link>{' '}
            avoid {type}.
          </Trans>
        }
      />
    </>
  )
}

      ↓ ↓ ↓ ↓ ↓ ↓

import React from 'react';
import { useTranslation, Trans } from 'react-i18next';

const Link = ({
  to,
  children
}) => <a href={to}>{children}</a>;

export default function TestPage({
  count = 1
}) {
  const [t] = useTranslation();
  const catchDate = Date.now();
  const completion = 0.75;
  const gender = Math.random() < 0.5 ? 'female' : 'male';
  return <>
      {t('sample.text', 'Some sample text with {word} {gender} {count, number} {catchDate, date} {completion, number, percent}', {
      word: 'interpolation',
      gender,
      count,
      catchDate,
      completion
    })}
      <Trans i18nKey=\\"plural\\" count={count} defaults=\\"{count, plural,  =0 {<0>Your cart</0> is <1>empty</1>.} one {You have <2># item</2> in <3>your cart</3>.} other {You have <4># items</4> in <5>your cart</5>.}}\\" components={[<Link to='/cart'>Your cart</Link>, <strong>empty</strong>, <strong># item</strong>, <Link to='/cart'>your cart</Link>, <strong># items</strong>, <Link to='/cart'>your cart</Link>]} values={{
      linkPath: \\"/item/\\" + count
    }} />
      <Trans i18nKey=\\"select\\" defaults=\\"{gender, select,  female {These are <0>her items</0>} male {These are <1>his items</1>} other {These are <2>their items</2>}}\\" components={[<Link to='/items'>her items</Link>, <Link to='/items'>his items</Link>, <Link to='/items'>their items</Link>]} values={{
      gender
    }} />
      <Trans i18nKey=\\"ordinal\\" count={itemIndex + 1} defaults=\\"{count, selectordinal,  one {Your <0><0>#st</0> item</0>} two {Your <1><0>#nd</0> item</1>} few {Your <2><0>#rd</0> item</2>} other {Your <3><0>#th</0> item</3>}}\\" components={[<Link to={linkPath}><strong>#st</strong> item</Link>, <Link to={linkPath}><strong>#nd</strong> item</Link>, <Link to={linkPath}><strong>#rd</strong> item</Link>, <Link to={linkPath}><strong>#th</strong> item</Link>]} values={{
      linkPath: \\"/item/\\" + itemIndex
    }} />
      <Trans i18nKey=\\"percent\\" defaults=\\"You've completed <0>{completion, number, percent} of your tasks</0>.\\" components={[<Link to='/tasks'>{(completion)} of your tasks</Link>]} values={{
      completion
    }} />
      <Trans i18nKey=\\"date\\" defaults=\\"Caught on <0>{catchDate, date, short}</0>!\\" components={[<Link to='/dest'>{(catchDate)}</Link>]} values={{
      catchDate
    }} />
      <Trans i18nKey=\\"ordinal.prettier\\" count={count} defaults=\\"{count, selectordinal,  one {Your <0><0>#st</0> {type}</0>} two {Your <1><0>#nd</0> {type}</1>} few {Your <2><0>#rd</0> {type}</2>} other {Your <3><0>#th</0> {type}</3>}}\\" components={[<Link to={linkPath}>
              <strong>#st</strong> {type}
            </Link>, <Link to={linkPath}>
              <strong>#nd</strong> {type}
            </Link>, <Link to={linkPath}>
              <strong>#rd</strong> {type}
            </Link>, <Link to={linkPath}>
              <strong>#th</strong> {type}
            </Link>]} values={{
      linkPath: \`/item/\${count}\`,
      type: 'item',
      prop
    }} />
      <Trans i18nKey=\\"select.expr.prettier\\" defaults=\\"{selectKey, select,  malePerson {<0><0>He</0></0> avoids {type}.} femalePerson {<1><0>She</0></1> avoids {type}.} other {<2><0>They</0></2> avoid {type}.}}\\" components={[<Link to={linkPath}>
              <strong>He</strong>
            </Link>, <Link to={linkPath}>
              <strong>She</strong>
            </Link>, <Link to={linkPath}>
              <strong>They</strong>
            </Link>]} values={{
      selectKey: \`\${gender}Person\`,
      linkPath: \`/users/\${number}\`,
      type: 'bugs'
    }} />
    </>;
}
"
`;

exports[`macros 21. macros: 21. macros 1`] = `
"
import React from \\"react\\"
import { Trans, number, date, time, plural, select, selectOrdinal } from \\"../icu.macro\\";

function Component({ children, style }) {
  return <div style={style}>{children}</div>
}

const count = 2;
const numbers = 34;
const selectInput = \\"thing\\"
const now = new Date()
const x = (
  <Trans i18nKey=\\"key\\">
    <strong>exciting!</strong>
    {plural\`\${count},
    =0 { hi there \${<strong>friend</strong>} }
    other { woweee even supports nested \${number\`\${numbers}\`} } \`} and
    {select\`\${selectInput},
     thing { another nested \${<Component style={{ color: \\"red\\" }}>
       with regular text and a date: <pre>{date\`\${now}\`}</pre>
     </Component>}} \`}
  </Trans>
);

      ↓ ↓ ↓ ↓ ↓ ↓

import { Trans, number, date, plural, select } from \\"react-i18next\\";
import React from \\"react\\";

function Component({
  children,
  style
}) {
  return <div style={style}>{children}</div>;
}

const count = 2;
const numbers = 34;
const selectInput = \\"thing\\";
const now = new Date();
const x = <Trans i18nKey=\\"key\\" defaults=\\"<0>exciting!</0>{count, plural, =0 { hi there <1>friend</1> } other { woweee even supports nested {numbers, number} } } and{selectInput, select, thing { another nested <2>with regular text and a date: <0>{now, date}</0></2>} }\\" components={[<strong>exciting!</strong>, <strong>friend</strong>, <Component style={{
  color: \\"red\\"
}}>
       with regular text and a date: <pre>{date\`\${now}\`}</pre>
     </Component>]} values={{
  count,
  numbers,
  selectInput,
  now
}} />;
"
`;
